controllers/azuremachine_reconciler.go

```diff
@@ -84,6 +88,11 @@ func (s *azureMachineService) Reconcile(ctx context.Context) (*infrav1.VM, error
 		return nil, errors.Wrapf(vmErr, "failed to create VM %s ", s.machineScope.Name())
 	}
 
+	err = s.roleAssignmentsSvc.Reconcile(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create role assignment")
+	}
+
```

Comment:
Having this role Assignment created in a separate step after the VM is already created can cause issues if it fails, the VM would be there with the wrong identity

right?

Reply:
yes you're right although the same thing could happen in the current code https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/811/files#diff-b2813f64759b38fd374267409eaaca65L193 since the VM gets created before the role assignment, if the role assignment fails to create then it would requeue and at the next reconcile it would get the VM, see that it exists and not reconcile.

I think the way we fix that is by removing https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b0e2d89ea5e941c114713fca837520e64266676b/controllers/azuremachine_controller.go#L340 once we switch the vm service to not be a getter service and instead at the beginning of the VM reconcile Get to VM and return early if it exists, but still reconcile the role assignment separately (hence the question of dealing with a requeue where role assignment was already created).

Reply:
ok makes sense